app/src/main/java/com/google/android/sambadocumentsprovider/encryption/EncryptionManager.java

```diff
@@ -0,0 +1,195 @@
+      if (key == null) {
+        throw new EncryptionException("No such encryption key: " + keyAlias);
+      }
+
+      Cipher cipher = Cipher.getInstance(AES_CIPHER);
+      cipher.init(Cipher.DECRYPT_MODE, key.mKey, new GCMParameterSpec(GCM_TAG_LENGTH, key.mIv));
+
+      return cipher.doFinal(data);
+    } catch (Exception e) {
+      throw new EncryptionException("Failed to decrypt data: ", e);
+    }
+  }
+
+  public void registerKey(String keyAlias) throws EncryptionException {
+    if (isKeyRegistered(keyAlias)) {
+      return;
+    }
+
+    try {
+      KeyGenerator keyGen = KeyGenerator.getInstance(
+              KeyProperties.KEY_ALGORITHM_AES, KEYSTORE_PROVIDER);
+
+      KeyGenParameterSpec spec = new KeyGenParameterSpec.Builder(
+              keyAlias, KeyProperties.PURPOSE_ENCRYPT | KeyProperties.PURPOSE_DECRYPT)
+              .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
+              .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
+              .setRandomizedEncryptionRequired(false)
+              .build();
+      keyGen.init(spec);
+
+      SecretKey key = keyGen.generateKey();
+      byte[] iv = generateIv();
+
+      mKeys.put(keyAlias, new EncryptionKey(key, iv));
+      saveIvForKey(keyAlias, iv);
+    } catch (NoSuchProviderException
+            | InvalidAlgorithmParameterException | NoSuchAlgorithmException e) {
+      throw new EncryptionException("Failed to register key: ", e);
+    }
+  }
+
+  public boolean isKeyRegistered(String keyAlias) {
+    return mKeys.containsKey(keyAlias);
+  }
+
+  private void loadKeyStore() throws KeyStoreException, CertificateException, NoSuchAlgorithmException, IOException {
+    mStore = KeyStore.getInstance(KEYSTORE_PROVIDER);
+    mStore.load(null);
+  }
+
+  private void loadKeys() throws KeyStoreException {
+    Enumeration<String> keyAliases = mStore.aliases();
+
+    while (keyAliases.hasMoreElements()) {
+      String alias = keyAliases.nextElement();
+
+      SecretKey key;
+      try {
+        key = (SecretKey) mStore.getKey(alias, null);
+      } catch (NoSuchAlgorithmException | UnrecoverableKeyException e) {
+        Log.e(TAG, "Failed to load encryption key: ", e);
+        continue;
+      }
+
+      EncryptionKey encryptionKey = new EncryptionKey(key, loadIvForKey(alias));
+      mKeys.put(alias, encryptionKey);
+    }
+  }
+
+  private static byte[] generateIv() {
+    byte[] iv = new byte[IV_LENGTH];
+
+    mRandom.nextBytes(iv);
+
+    return iv;
+  }
+
+  private byte[] loadIvForKey(String alias) {
+    String data = mPref.getString(alias, null);
+    if (data == null) {
+      byte[] iv = generateIv();
+      saveIvForKey(alias, iv);
+      return iv;
+    }
+
+    return Base64.decode(data, Base64.DEFAULT);
+  }
+
+  private void saveIvForKey(String alias, byte[] iv) {
+    String data = Base64.encodeToString(iv, Base64.DEFAULT);
+    mPref.edit().putString(alias, data).apply();
+  }
+
+  private static class EncryptionKey {
```

Comment:
nit: Even though this is a private inner class, still mark fields private and use getters.